File: ios/liquid_ai/Sources/liquid_ai/ConversationManager.swift
import Foundation
import LeapSDK

/// Manages conversations and active generations.
actor ConversationManager {
    /// Stored conversations keyed by conversation ID.
    private var conversations: [String: ConversationState] = [:]

    /// Active generation tasks keyed by generation ID.
    private var activeGenerations: [String: Task<Void, Never>] = [:]

    /// Cancelled generation IDs.
    private var cancelledGenerations: Set<String> = []

    /// Progress handler for streaming events to Flutter.
    private let progressHandler: GenerationProgressHandler

    /// Reference to the model runner manager.
    private let runnerManager: ModelRunnerManager

    init(progressHandler: GenerationProgressHandler, runnerManager: ModelRunnerManager) {
        self.progressHandler = progressHandler
        self.runnerManager = runnerManager
    }

    // MARK: - Conversation Lifecycle

    /// Creates a new conversation.
    func createConversation(
        runnerId: String,
        systemPrompt: String?
    ) async throws -> String {
        guard let runner = await runnerManager.getRunner(runnerId: runnerId) else {
            throw ConversationError.runnerNotFound(runnerId)
        }

        let conversationId = UUID().uuidString

        var history: [ChatMessage] = []
        if let systemPrompt = systemPrompt {
            history.append(ChatMessage(
                role: .system,
                content: [.text(systemPrompt)]
            ))
        }

        let conversation = Conversation(modelRunner: runner, history: history)

        let state = ConversationState(
            runnerId: runnerId,
            conversation: conversation,
            history: history
        )

        conversations[conversationId] = state
        return conversationId
    }

    /// Creates a conversation from existing history.
    func createConversationFromHistory(
        runnerId: String,
        history: [[String: Any]]
    ) async throws -> String {
        guard let runner = await runnerManager.getRunner(runnerId: runnerId) else {
            throw ConversationError.runnerNotFound(runnerId)
        }

        let conversationId = UUID().uuidString
        let messages = history.compactMap { Self.parseChatMessage($0) }
        let conversation = Conversation(modelRunner: runner, history: messages)

        let state = ConversationState(
            runnerId: runnerId,
            conversation: conversation,
            history: messages
        )

        conversations[conversationId] = state
        return conversationId
    }

    /// Gets the conversation history.
    func getHistory(conversationId: String) throws -> [[String: Any]] {
        guard let state = conversations[conversationId] else {
            throw ConversationError.conversationNotFound(conversationId)
        }

        return state.history.map { Self.serializeChatMessage($0) }
    }

    /// Disposes of a conversation.
    func disposeConversation(_ conversationId: String) {
        conversations.removeValue(forKey: conversationId)
    }

    /// Exports a conversation as JSON.
    func exportConversation(_ conversationId: String) throws -> String {
        guard let state = conversations[conversationId] else {
            throw ConversationError.conversationNotFound(conversationId)
        }

        let export: [String: Any] = [
            "conversationId": conversationId,
            "runnerId": state.runnerId,
            "messages": state.history.map { Self.serializeChatMessage($0) }
        ]

        let data = try JSONSerialization.data(withJSONObject: export, options: .prettyPrinted)
        return String(data: data, encoding: .utf8) ?? "{}"
    }

    // MARK: - Generation

    /// Generates a response to the given message.
    func generateResponse(
        conversationId: String,
        message: [String: Any],
        options: [String: Any]?
    ) async throws -> String {
        guard var state = conversations[conversationId] else {
            throw ConversationError.conversationNotFound(conversationId)
        }

        guard let userMessage = Self.parseChatMessage(message) else {
            throw ConversationError.invalidMessage
        }

        let generationId = UUID().uuidString
        cancelledGenerations.remove(generationId)

        // Add user message to history
        state.history.append(userMessage)
        conversations[conversationId] = state

        let task = Task { [weak self] in
            guard let self = self else { return }

            do {
                var generatedText = ""
                var tokenCount = 0
                let startTime = Date()

                // Get the conversation from state
                guard let currentState = await self.getConversationState(conversationId) else {
                    return
                }

                // Use the SDK's conversation.generateResponse
                for try await response in currentState.conversation.generateResponse(message: userMessage) {
                    let isCancelled = await self.isGenerationCancelled(generationId)
                    if isCancelled {
                        self.progressHandler.sendCancelled(generationId: generationId)
                        await self.removeGeneration(generationId)
                        return
                    }

                    switch response {
                    case .chunk(let text):
                        generatedText += text
                        tokenCount += 1
                        self.progressHandler.sendChunk(generationId: generationId, chunk: text)

                    case .reasoningChunk(let reasoning):
                        self.progressHandler.sendReasoningChunk(generationId: generationId, chunk: reasoning)

                    case .audioSample(let samples, let sampleRate):
                        self.progressHandler.sendAudioSamples(
                            generationId: generationId,
                            samples: samples,
                            sampleRate: sampleRate
                        )

                    case .functionCall(let calls):
                        let serializedCalls = calls.enumerated().map { (index, call) -> [String: Any] in
                            [
                                "id": "call_\(index)",
                                "name": call.name,
                                "arguments": call.arguments as Any
                            ]
                        }
                        self.progressHandler.sendFunctionCalls(
                            generationId: generationId,
                            calls: serializedCalls
                        )

                    case .complete(let completion):
                        let isCancelled = await self.isGenerationCancelled(generationId)
                        if isCancelled {
                            self.progressHandler.sendCancelled(generationId: generationId)
                            await self.removeGeneration(generationId)
                            return
                        }

                        // Create assistant message from completion
                        let assistantMessage = completion.message

                        // Add to history
                        await self.appendMessage(conversationId: conversationId, message: assistantMessage)

                        let duration = Date().timeIntervalSince(startTime)
                        let tokensPerSecond = duration > 0 ? Double(tokenCount) / duration : 0

                        let finishReason = Self.mapFinishReason(completion.finishReason)

                        var stats: [String: Any] = [
                            "tokenCount": tokenCount,
                            "tokensPerSecond": tokensPerSecond,
                            "generationTimeMs": Int(duration * 1000)
                        ]

                        if let genStats = completion.stats {
                            stats["promptTokenCount"] = genStats.promptTokens
                            stats["completionTokenCount"] = genStats.completionTokens
                        }

                        self.progressHandler.sendComplete(
                            generationId: generationId,
                            message: Self.serializeChatMessage(assistantMessage),
                            finishReason: finishReason,
                            stats: stats
                        )

                        await self.removeGeneration(generationId)
                    }
                }

            } catch {
                let isCancelled = await self.isGenerationCancelled(generationId)
                if !isCancelled {
                    self.progressHandler.sendError(
                        generationId: generationId,
                        error: error.localizedDescription
                    )
                }
                await self.removeGeneration(generationId)
            }
        }

        activeGenerations[generationId] = task
        return generationId
    }

    /// Stops an ongoing generation.
    func stopGeneration(_ generationId: String) {
        cancelledGenerations.insert(generationId)
        activeGenerations[generationId]?.cancel()
        activeGenerations.removeValue(forKey: generationId)
        progressHandler.sendCancelled(generationId: generationId)
    }

    // MARK: - Function Calling

    /// Registers a function for a conversation.
    func registerFunction(
        conversationId: String,
        function: [String: Any]
    ) throws {
        guard conversations[conversationId] != nil else {
            throw ConversationError.conversationNotFound(conversationId)
        }

        // Function calling support would be implemented here
        // when the LEAP SDK adds function calling capabilities
    }

    /// Provides a function result back to the conversation.
    func provideFunctionResult(
        conversationId: String,
        result: [String: Any]
    ) throws {
        guard var state = conversations[conversationId] else {
            throw ConversationError.conversationNotFound(conversationId)
        }

        if let callId = result["callId"] as? String,
           let resultText = result["result"] as? String {
            // Add function result as a message
            let message = ChatMessage(
                role: .user,
                content: [.text("Function call \(callId) result: \(resultText)")]
            )
            state.history.append(message)
            conversations[conversationId] = state
        }
    }

    // MARK: - Private Helpers

    private func getConversationState(_ conversationId: String) -> ConversationState? {
        return conversations[conversationId]
    }

    private func appendMessage(conversationId: String, message: ChatMessage) {
        if var state = conversations[conversationId] {
            state.history.append(message)
            conversations[conversationId] = state
        }
    }

    private func isGenerationCancelled(_ generationId: String) -> Bool {
        return cancelledGenerations.contains(generationId)
    }

    private func removeGeneration(_ generationId: String) {
        activeGenerations.removeValue(forKey: generationId)
    }

    // MARK: - Serialization Helpers

    static func parseChatMessage(_ map: [String: Any]) -> ChatMessage? {
        guard let roleString = map["role"] as? String,
              let contentList = map["content"] as? [[String: Any]] else {
            return nil
        }

        let role: ChatMessageRole
        switch roleString {
        case "system": role = .system
        case "user": role = .user
        case "assistant": role = .assistant
        default: return nil
        }

        var content: [ChatMessageContent] = []
        for item in contentList {
            guard let type = item["type"] as? String else { continue }
            switch type {
            case "text":
                if let text = item["text"] as? String {
                    content.append(.text(text))
                }
            case "image":
                if let dataArray = item["data"] as? [Int] {
                    let data = Data(dataArray.map { UInt8($0) })
                    content.append(.image(data))
                }
            case "audio":
                if let dataArray = item["data"] as? [Int] {
                    let data = Data(dataArray.map { UInt8($0) })
                    content.append(.audio(data))
                }
            default:
                break
            }
        }

        return ChatMessage(role: role, content: content)
    }

    static func serializeChatMessage(_ message: ChatMessage) -> [String: Any] {
        let roleString: String
        switch message.role {
        case .system: roleString = "system"
        case .user: roleString = "user"
        case .assistant: roleString = "assistant"
        @unknown default: roleString = "user"
        }

        var contentList: [[String: Any]] = []
        for item in message.content {
            switch item {
            case .text(let text):
                contentList.append(["type": "text", "text": text])
            case .image(let data):
                contentList.append(["type": "image", "data": Array(data).map { Int($0) }])
            case .audio(let data):
                contentList.append(["type": "audio", "data": Array(data).map { Int($0) }])
            @unknown default:
                break
            }
        }

        return [
            "role": roleString,
            "content": contentList
        ]
    }

    static func mapFinishReason(_ reason: GenerationFinishReason) -> String {
        switch reason {
        case .stop:
            return "endOfSequence"
        case .exceed_context:
            return "maxTokens"
        @unknown default:
            return "endOfSequence"
        }
    }
}

// MARK: - Supporting Types

/// State for a single conversation.
struct ConversationState {
    let runnerId: String
    let conversation: Conversation
    var history: [ChatMessage]
}

/// Errors that can occur during conversation management.
enum ConversationError: LocalizedError {
    case runnerNotFound(String)
    case conversationNotFound(String)
    case invalidMessage
    case generationFailed(String)

    var errorDescription: String? {
        switch self {
        case .runnerNotFound(let id):
            return "Model runner not found: \(id)"
        case .conversationNotFound(let id):
            return "Conversation not found: \(id)"
        case .invalidMessage:
            return "Invalid message format"
        case .generationFailed(let reason):
            return "Generation failed: \(reason)"
        }
    }
}

File: android/src/main/kotlin/dev/sogl/liquid_ai/ConversationManager.kt
package dev.sogl.liquid_ai

import ai.liquid.leap.Conversation
import ai.liquid.leap.GenerationOptions
import ai.liquid.leap.message.ChatMessage
import ai.liquid.leap.message.MessageResponse
import ai.liquid.leap.function.LeapFunction
import ai.liquid.leap.function.LeapFunctionParameter
import ai.liquid.leap.function.LeapFunctionParameterType
import ai.liquid.leap.function.HermesFunctionCallParser
import ai.liquid.leap.function.LFMFunctionCallParser
import ai.liquid.leap.gson.registerLeapAdapters
import com.google.gson.GsonBuilder
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.flow.onCompletion
import kotlinx.coroutines.flow.onEach
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

/// Manages conversations and active generations.
class ConversationManager(
    private val progressHandler: GenerationProgressHandler,
    private val runnerManager: ModelRunnerManager
) {
    private val conversations = ConcurrentHashMap<String, ConversationState>()
    private val activeGenerations = ConcurrentHashMap<String, Job>()
    private val cancelledGenerations = ConcurrentHashMap.newKeySet<String>()
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    /// Gson instance with Leap SDK adapters for ChatMessage serialization.
    private val gson = GsonBuilder().registerLeapAdapters().create()

    // MARK: - Conversation Lifecycle

    /// Creates a new conversation.
    fun createConversation(
        runnerId: String,
        systemPrompt: String?
    ): String {
        val runner = runnerManager.getRunner(runnerId)
            ?: throw ConversationException("Model runner not found: $runnerId")

        val conversationId = UUID.randomUUID().toString()

        // Create SDK conversation with optional system prompt
        val conversation = if (systemPrompt != null) {
            runner.createConversation(systemPrompt)
        } else {
            runner.createConversation()
        }

        val history = mutableListOf<ChatMessage>()
        if (systemPrompt != null) {
            history.add(ChatMessage(
                role = ChatMessage.Role.SYSTEM,
                textContent = systemPrompt
            ))
        }

        val state = ConversationState(
            runnerId = runnerId,
            conversation = conversation,
            history = history
        )

        conversations[conversationId] = state
        return conversationId
    }

    /// Creates a conversation from existing history.
    fun createConversationFromHistory(
        runnerId: String,
        history: List<Map<String, Any>>
    ): String {
        val runner = runnerManager.getRunner(runnerId)
            ?: throw ConversationException("Model runner not found: $runnerId")

        val conversationId = UUID.randomUUID().toString()
        val messages = history.mapNotNull { parseFlutterMessageToSdk(it) }.toMutableList()

        // Create conversation - SDK may support history initialization
        val conversation = runner.createConversation()

        val state = ConversationState(
            runnerId = runnerId,
            conversation = conversation,
            history = messages
        )

        conversations[conversationId] = state
        return conversationId
    }

    /// Gets the conversation history.
    fun getHistory(conversationId: String): List<Map<String, Any>> {
        val state = conversations[conversationId]
            ?: throw ConversationException("Conversation not found: $conversationId")

        return state.history.map { serializeSdkMessageToFlutter(it) }
    }

    /// Disposes of a conversation.
    fun disposeConversation(conversationId: String) {
        conversations.remove(conversationId)
    }

    /// Exports a conversation as JSON using Gson with Leap SDK adapters.
    fun exportConversation(conversationId: String): String {
        val state = conversations[conversationId]
            ?: throw ConversationException("Conversation not found: $conversationId")

        val export = mapOf(
            "conversationId" to conversationId,
            "runnerId" to state.runnerId,
            "messages" to state.history
        )

        return gson.toJson(export)
    }

    // MARK: - Generation

    /// Generates a response to the given message.
    fun generateResponse(
        conversationId: String,
        message: Map<String, Any>,
        options: Map<String, Any>?
    ): String {
        val state = conversations[conversationId]
            ?: throw ConversationException("Conversation not found: $conversationId")

        val sdkMessage = parseFlutterMessageToSdk(message)
            ?: throw ConversationException("Invalid message format")

        val generationId = UUID.randomUUID().toString()
        cancelledGenerations.remove(generationId)

        // Add message to our history tracking
        state.history.add(sdkMessage)

        val job = scope.launch {
            try {
                val generatedText = StringBuilder()
                var tokenCount = 0
                val startTime = System.currentTimeMillis()
                var isComplete = false

                // Parse generation options from Flutter
                val generationOptions = parseGenerationOptions(options)

                // Use SDK's conversation.generateResponse with proper ChatMessage
                state.conversation.generateResponse(sdkMessage, generationOptions)
                    .onEach { response ->
                        if (cancelledGenerations.contains(generationId)) {
                            progressHandler.sendCancelled(generationId)
                            activeGenerations.remove(generationId)
                            return@onEach
                        }

                        when (response) {
                            is MessageResponse.Chunk -> {
                                generatedText.append(response.text)
                                tokenCount++
                                progressHandler.sendChunk(generationId, response.text)
                            }
                            is MessageResponse.ReasoningChunk -> {
                                progressHandler.sendReasoningChunk(generationId, response.reasoning)
                            }
                            is MessageResponse.FunctionCalls -> {
                                val serializedCalls = response.functionCalls.mapIndexed { index, call ->
                                    mapOf(
                                        "id" to "call_$index",
                                        "name" to call.name,
                                        "arguments" to call.arguments
                                    )
                                }
                                progressHandler.sendFunctionCalls(generationId, serializedCalls)
                            }
                            is MessageResponse.Complete -> {
                                isComplete = true
                                // Complete event handled in onCompletion
                            }
                            else -> {
                                // Handle other response types if any
                            }
                        }
                    }
                    .onCompletion { error ->
                        if (error == null && !cancelledGenerations.contains(generationId)) {
                            val assistantMessage = ChatMessage(
                                role = ChatMessage.Role.ASSISTANT,
                                textContent = generatedText.toString()
                            )

                            state.history.add(assistantMessage)

                            val duration = System.currentTimeMillis() - startTime
                            val tokensPerSecond = if (duration > 0) {
                                tokenCount.toDouble() / (duration.toDouble() / 1000.0)
                            } else 0.0

                            progressHandler.sendComplete(
                                generationId = generationId,
                                message = serializeSdkMessageToFlutter(assistantMessage),
                                finishReason = "stop",
                                stats = mapOf(
                                    "tokenCount" to tokenCount,
                                    "tokensPerSecond" to tokensPerSecond,
                                    "generationTimeMs" to duration.toInt()
                                )
                            )
                        }
                        activeGenerations.remove(generationId)
                    }
                    .catch { e ->
                        if (!cancelledGenerations.contains(generationId)) {
                            progressHandler.sendError(
                                generationId = generationId,
                                error = e.message ?: "Generation failed"
                            )
                        }
                        activeGenerations.remove(generationId)
                    }
                    .collect()

            } catch (e: Exception) {
                if (!cancelledGenerations.contains(generationId)) {
                    progressHandler.sendError(
                        generationId = generationId,
                        error = e.message ?: "Generation failed"
                    )
                }
                activeGenerations.remove(generationId)
            }
        }

        activeGenerations[generationId] = job
        return generationId
    }

    /// Stops an ongoing generation.
    fun stopGeneration(generationId: String) {
        cancelledGenerations.add(generationId)
        activeGenerations[generationId]?.cancel()
        activeGenerations.remove(generationId)
        progressHandler.sendCancelled(generationId)
    }

    // MARK: - Function Calling

    /// Registers a function for a conversation.
    @Suppress("UNCHECKED_CAST")
    fun registerFunction(
        conversationId: String,
        function: Map<String, Any>
    ) {
        val state = conversations[conversationId]
            ?: throw ConversationException("Conversation not found: $conversationId")

        // Parse the function definition from Flutter
        val name = function["name"] as? String
            ?: throw ConversationException("Function name is required")
        val description = function["description"] as? String
            ?: throw ConversationException("Function description is required")
        val parameters = function["parameters"] as? Map<String, Any>
            ?: throw ConversationException("Function parameters are required")

        // Parse parameters schema into LeapFunctionParameter list
        val leapParameters = parseParameters(parameters)

        // Create LeapFunction and register with conversation
        val leapFunction = LeapFunction(
            name = name,
            description = description,
            parameters = leapParameters
        )

        state.conversation.registerFunction(leapFunction)

        android.util.Log.d("LiquidAI", "Registered function: $name with ${leapParameters.size} parameters")
    }

    /// Provides a function result back to the conversation.
    ///
    /// This adds the function result as a user message to the conversation
    /// history so the model can continue the conversation with the result.
    fun provideFunctionResult(
        conversationId: String,
        result: Map<String, Any>
    ) {
        val state = conversations[conversationId]
            ?: throw ConversationException("Conversation not found: $conversationId")

        val callId = result["callId"] as? String
            ?: throw ConversationException("Call ID is required")
        val resultText = result["result"] as? String
            ?: throw ConversationException("Result text is required")

        // Add function result as a user message so the model can process it
        state.history.add(ChatMessage(
            role = ChatMessage.Role.USER,
            textContent = "Function call $callId result: $resultText"
        ))

        android.util.Log.d("LiquidAI", "Provided function result for call: $callId")
    }

    /// Cleans up resources.
    fun dispose() {
        scope.cancel()
        conversations.clear()
        activeGenerations.clear()
        cancelledGenerations.clear()
    }

    // MARK: - Generation Options Parsing

    /// Parses Flutter generation options into LeapSDK GenerationOptions.
    private fun parseGenerationOptions(options: Map<String, Any>?): GenerationOptions? {
        if (options == null) return null

        return GenerationOptions().apply {
            (options["temperature"] as? Double)?.let {
                temperature = it.toFloat()
            }
            (options["topP"] as? Double)?.let {
                topP = it.toFloat()
            }
            (options["minP"] as? Double)?.let {
                minP = it.toFloat()
            }
            (options["repetitionPenalty"] as? Double)?.let {
                repetitionPenalty = it.toFloat()
            }
            (options["maxTokens"] as? Int)?.let {
                maxOutputTokens = it.toUInt()
            }
            // JSON schema constraint for structured output
            (options["jsonSchemaConstraint"] as? String)?.let { schema ->
                android.util.Log.d("LiquidAI", "Setting jsonSchemaConstraint: ${schema.take(200)}...")
                jsonSchemaConstraint = schema
            }
            // Function call parser configuration
            (options["functionCallParser"] as? String)?.let { parserType ->
                functionCallParser = when (parserType) {
                    "hermes" -> {
                        android.util.Log.d("LiquidAI", "Using HermesFunctionCallParser")
                        HermesFunctionCallParser()
                    }
                    "raw", "none" -> {
                        android.util.Log.d("LiquidAI", "Using raw function call output (no parser)")
                        null
                    }
                    "lfm", "default" -> {
                        android.util.Log.d("LiquidAI", "Using LFMFunctionCallParser")
                        LFMFunctionCallParser()
                    }
                    else -> null // Keep default
                }
            }
        }
    }

    // MARK: - Function Parameter Parsing

    /// Parses a JSON Schema parameters object into LeapFunctionParameter list.
    @Suppress("UNCHECKED_CAST")
    private fun parseParameters(schema: Map<String, Any>): List<LeapFunctionParameter> {
        val properties = schema["properties"] as? Map<String, Map<String, Any>>
            ?: return emptyList()

        val requiredFields = (schema["required"] as? List<String>) ?: emptyList()

        return properties.mapNotNull { (name, propertySchema) ->
            val typeString = propertySchema["type"] as? String ?: return@mapNotNull null
            val description = propertySchema["description"] as? String ?: ""
            val isOptional = !requiredFields.contains(name)

            val parameterType = parseParameterType(typeString, propertySchema)

            LeapFunctionParameter(
                name = name,
                type = parameterType,
                description = description,
                optional = isOptional
            )
        }
    }

    /// Parses a JSON Schema type into LeapFunctionParameterType.
    @Suppress("UNCHECKED_CAST")
    private fun parseParameterType(
        typeString: String,
        schema: Map<String, Any>
    ): LeapFunctionParameterType {
        val enumValues = schema["enum"] as? List<String>

        return when (typeString) {
            "string" -> LeapFunctionParameterType.String(enumValues = enumValues)
            "number" -> LeapFunctionParameterType.Number()
            "integer" -> LeapFunctionParameterType.Integer()
            "boolean" -> LeapFunctionParameterType.Boolean()
            "array" -> {
                // Parse array item type if available
                val itemsSchema = schema["items"] as? Map<String, Any>
                val itemType = itemsSchema?.get("type") as? String
                if (itemType != null && itemsSchema != null) {
                    val itemParamType = parseParameterType(itemType, itemsSchema)
                    LeapFunctionParameterType.Array(itemType = itemParamType)
                } else {
                    LeapFunctionParameterType.Array(itemType = LeapFunctionParameterType.String())
                }
            }
            "object" -> {
                // Parse nested object properties
                val nestedProps = schema["properties"] as? Map<String, Map<String, Any>>
                if (nestedProps != null) {
                    val properties = nestedProps.mapNotNull { (propName, propSchema) ->
                        val propType = propSchema["type"] as? String ?: return@mapNotNull null
                        propName to parseParameterType(propType, propSchema)
                    }.toMap()
                    val required = (schema["required"] as? List<String>) ?: emptyList()
                    LeapFunctionParameterType.Object(properties = properties, required = required)
                } else {
                    LeapFunctionParameterType.Object(properties = emptyMap(), required = emptyList())
                }
            }
            else -> LeapFunctionParameterType.String()
        }
    }

    // MARK: - Private Helpers

    /// Parses a Flutter message map into an SDK ChatMessage.
    @Suppress("UNCHECKED_CAST")
    private fun parseFlutterMessageToSdk(map: Map<String, Any>): ChatMessage? {
        val roleString = map["role"] as? String ?: return null
        val contentList = map["content"] as? List<Map<String, Any>> ?: return null

        // Map role string to SDK role enum
        val role = when (roleString) {
            "system" -> ChatMessage.Role.SYSTEM
            "user" -> ChatMessage.Role.USER
            "assistant" -> ChatMessage.Role.ASSISTANT
            "tool" -> ChatMessage.Role.TOOL
            else -> ChatMessage.Role.USER
        }

        // Extract text content (primary use case)
        val textContent = contentList
            .filter { it["type"] == "text" }
            .mapNotNull { it["text"] as? String }
            .joinToString(" ")

        return ChatMessage(role = role, textContent = textContent)
    }

    /// Serializes an SDK ChatMessage to a Flutter-compatible map.
    private fun serializeSdkMessageToFlutter(message: ChatMessage): Map<String, Any> {
        val roleString = when (message.role) {
            ChatMessage.Role.SYSTEM -> "system"
            ChatMessage.Role.USER -> "user"
            ChatMessage.Role.ASSISTANT -> "assistant"
            ChatMessage.Role.TOOL -> "tool"
        }

        // Build content list from SDK message
        val contentList = mutableListOf<Map<String, Any>>()

        // Add text content if present
        message.textContent?.let { text ->
            if (text.isNotEmpty()) {
                contentList.add(mapOf("type" to "text", "text" to text))
            }
        }

        return mapOf(
            "role" to roleString,
            "content" to contentList
        )
    }
}

// MARK: - Supporting Types

/// Internal representation of conversation state.
data class ConversationState(
    val runnerId: String,
    val conversation: Conversation,
    val history: MutableList<ChatMessage>
)

/// Exception for conversation errors.
class ConversationException(message: String) : Exception(message)
